---
title: "Using admixtools in R"
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
---


For the examples below the following R packages need to be loaded.

```{r, results='hide', message=FALSE, warning=FALSE}
library(admixtools)
library(magrittr)
library(tidyverse)
```

If you have access to pre-computed f2-statistics, you can use them to compute f3 and f4 statistics, admixture weights, and to fit admixture graphs. The following examples should work on O2. The examples are just there to illustrate the principle, the models themselves don't make too much sense.

<br>

## qp3Pop

```{r}
pop1 = 'Denisova.DG'
pop2 = c('Altai_Neanderthal.DG', 'Vindija.DG')
pop3 = c('Chimp.REF', 'Mbuti.DG', 'Russia_Ust_Ishim.DG')
```
```{r, eval = FALSE}
f2_dir = '/n/groups/reich/robert/projects/admixprograms/f2blocks_v42.1/'
qp3pop(pop1, pop2, pop3, f2_dir = f2_dir)
```
```{r, echo = FALSE}
qp3pop(pop1, pop2, pop3, example_f2_blocks, example_block_lengths)
```


<br>

## qpDstat

```{r}
pop4 = 'Switzerland_Bichon.SG'
```
```{r, eval = FALSE}
qpdstat(pop1, pop2, pop3, pop4, f2_dir = f2_dir)
```
```{r, echo = FALSE}
qpdstat(pop1, pop2, pop3, pop4, example_f2_blocks, example_block_lengths)
```


<br>

## qpAdm

```{r}
target = 'Denisova.DG'
left = c('Altai_Neanderthal.DG', 'Vindija.DG')
right = c('Chimp.REF', 'Mbuti.DG', 'Russia_Ust_Ishim.DG', 'Switzerland_Bichon.SG')
```
```{r, eval = FALSE}
qpadm(target, left, right, f2_dir = f2_dir)
```
```{r, echo = FALSE}
qpadm(target, left, right, example_f2_blocks, example_block_lengths)
```

### Caveats

This function only reports `qpAdm` admixture weights, and doesn't determine the number of independent source populations. The `qpadm` function also hasn't been tested as much as the `qpgraph` function, and the results are sometimes not very accurate, in particular when the weights are outside the range of [0, 1], like in this example.

<br>

## qpGraph


The f2-statistics can also be used to fit an admixture graph to the data.

```{r, eval = FALSE}
qpg_results = qpgraph(example_graph, f2_dir = f2_dir, f2_denom = 0.278)
```
```{r, echo = FALSE}
qpg_results = qpgraph(example_graph, example_f2_blocks, example_block_lengths, f2_denom = 0.278)
```

Here, `example_graph` is a specific graph included in this R package, but you can provide any other graph in one of three formats.

1. An [`igraph`](https://igraph.org/r/) object.
2. A two column matrix, where each row is an edge, the first column is the source of the edge, and the second column is the target.
3. The location of a `qpGraph` graph file, which will be read and parsed.

The only requirements for the input graph are that the graph is a valid admixture graph -- a DAG with one outgroup edge connecting the root to a leaf node -- and that f2-statistics exist for all leaf nodes. The first edge should go from the root to the outgroup.

<br>

The precomputed f2-statistics are not scaled, which it why it is necessary to set `f2_denom` to `0.278` to get matching results. Without that scaling, the absolute values may differ, but the relative values should still be ok.

<br>

The `qpgraph` output is a list with five items:

1. A data frame with estimated edge weights. This includes normal edges as well as admixture edges
2. The likelihood score. Smaller scores indicate a better fit
3. The estimated f2-statistics and standard errors for all population pairs
4. The estimated f3-statistics and standard errors for all population pairs with the outgroup
5. A data frame with details from the weight optimization, with one row per set of starting values


```{r}
qpg_results$edges
```

<br>

This works for any populations for which there are precomputed f2-statistics.

```{r, eval = FALSE}
allpops = list.dirs(f2_dir, full.names = FALSE, recursive = FALSE)
rand_pops = sample(allpops, 20)
rand_graph = random_admixturegraph(rand_pops, numadmix = 3)
rand_results = qpgraph(rand_graph, f2_dir = f2_dir, f2_denom = 0.278, verbose = TRUE)
```

### Caveats

* A 'fudge' factor is added to the diagonal elements of the covariance matrix, to ensure that the matrix inversion is stable. This factor defaults to a relatively large value of 1e-3, to ensure that no error will be thrown in most cases. However, the original qpGraph uses 1e-5, which gives more accurate results in most cases. Increasing this factor seems to have a small effect on admixture weights and branch lengths, but a larger effect on the likelihood scores. 
* There may be differences in the results that originate in how missing data is treated. The precomputed f2 statistics on O2 use all SNPs, similar to `useallsnps YES`, while the examples here were computed on f2 statistics based only on SNPs with non-missing data in all populations (`na.action = 'remove'`), similar to `useallsnps NO`.


<br>


## Comparing results to AdmixTools

There are wrapper function which call the origial `AdmixTools` programs and read the results. This can be used to check if the results of `qpgraph` are correct.

```{r, eval = FALSE}
binpath = '/home/np29/o2bin/'
env = 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/n/app/openblas/0.2.19/lib/:/n/app/gsl/2.3/lib/;'

qp3pop_bin  = paste0(env, binpath, 'qp3pop')
qpdstat_bin = paste0(env, binpath, 'qpDstat')
qpadm_bin   = paste0(env, binpath, 'qpAdm')
qpgraph_bin = paste0(env, binpath, 'qpGraph')

#prefix = '/n/groups/reich/DAVID/V42/V42.1/v42.1'
prefix = '/n/groups/reich/robert/projects/admixprograms/v42.1_small'
outdir = 'write/files/here/'

qp3pop_wrapper(source1 = pop2, source2 = pop3, target = pop1,
               qp3pop_bin, prefix, outdir)
qpdstat_wrapper(pop1, pop2, pop3, pop4,
                qpdstat_bin, prefix, outdir)
qpadm_wrapper(target, left, right,
              qpadm_bin, prefix, outdir)
qpgraph_wrapper(example_graph,
                qpgraph_bin, prefix, outdir)
```
**Unless `outdir` is specified, calling these wrapper functions may overwrite files in the working directory!**

If you already have existing parameter files and population or graph files, you can run the wrapper functions like this (though the different programs will require different parfiles and popfiles):

```{r, eval = FALSE}
qp3pop_wrapper (parfile = 'parfile.txt', bin = qp3pop_bin)
qpdstat_wrapper(parfile = 'parfile.txt', bin = qpdstat_bin)
qpadm_wrapper  (parfile = 'parfile.txt', bin = qpadm_bin)

qp3pop_wrapper ('popfile.txt', bin = qp3pop_bin,  pref = prefix)
qpdstat_wrapper('popfile.txt', bin = qpdstat_bin, pref = prefix)

qpadm_wrapper(left = 'left.txt', right = 'right.txt', bin = qpadm_bin, pref = prefix)

qpgraph_wrapper('graphfile.txt', parfile = 'parfile.txt', bin = qpgraph_bin)
```


<br>

The following function makes it easy to compare qpGraph results.

```{r, eval = FALSE}
qpgraph_ref_results = qpgraph_wrapper(example_graph, qpgraph_bin, prefix)
```
```{r, echo = FALSE}
qpgraph_ref_results = example_qpgraph_ref_results
```

```{r}
plot_comparison(qpg_results, qpgraph_ref_results)
```
<br>

There are at least a few examples where the results seem to be approximately correct.

**Please let me know if you find examples where the results are different!**



<br>

## Computing f2-statistics

If you're not on O2, or want use different samples or populations, you first have to compute f2 block jackknife statistics from genotype data.


```{r, eval = FALSE}
my_f2_dir = 'store/f2data/here/'
pops = c('Denisova.DG',
         'Altai_Neanderthal.DG',
         'Vindija.DG',
         'Chimp.REF',
         'Mbuti.DG',
         'Russia_Ust_Ishim.DG',
         'Switzerland_Bichon.SG')

extract_data(prefix, outdir = my_f2_dir, pops = pops)
```
This will look for genotype files in `packedancestrymap` or `PLINK` format, compute allele frequencies and f2 block jackknife statistics, and write them to `outdir`. Specifying `inds` rather than `pops` will treat each individual as a separate population. This is not very useful at the moment, because there is no simple way of combining block jackknife f2 statistics computed per individual to population block jackknife f2 statistics, but it may be useful in the future.

<br>

Now you should be able to run the same examples as before using your own f2 statistics.

```{r, eval = FALSE}
qpg_results = qpgraph(example_graph, f2_dir = my_f2_dir)
```
```{r, fig.width = 9, fig.align="center"}
plot_graph(qpg_results$edges)
```


The example above will read the f2 data from `my_f2_dir`. It can be faster to load it into memory first, like so:

```{r, eval = FALSE}
f2_blocks = read_f2(my_f2_dir, pops)
block_lengths = read_bl(my_f2_dir)

qpg_results = qpgraph(example_graph, f2_blocks, block_lengths)
```

<br>

You can also skip writing data to disk altogether, by not specifying `outdir` in `extract_data`. This can be useful if the number of populations is not too large:

```{r, eval = FALSE}
data = extract_data(prefix, pops = pops)
f2_blocks = data$f2_blocks
block_lengths = data$block_lengths
```


<br>

## Finding graphs with good fits

Since evaluating one graph doesn't take long once the f-statistics are pre-computed, we can evaluate many graphs and find graph topologies with good fits.

<br>

In order for that to be as fast as possible, we can precompute and reuse f3-statistics and the inverse of the f3-statistic covariance matrix.

```{r, eval = FALSE}
precomp = qpgraph_precompute_f3(pops, f2_dir = my_f2_dir, f2_denom = 0.278)
f3_jest = precomp$f3_jest
ppinv = precomp$ppinv
```
<br>

The function `optimize_admixturegraph` generates and evaluates `numgraphs` admixture graphs in `numgen` iterations across `numrep` independent repeats to find well fitting admixture graphs.

```{r, eval = FALSE}
opt_results = optimize_admixturegraph(pops, f3_jest, ppinv,
                                      numrep = 200, numgraphs = 100, numgen = 20,
                                      numsel = 5, numadmix = 3)
```
<br>

Once that is done, the following commands will extract the best fitting model overall, and the best fitting model from each independent repeat.

```{r, eval = FALSE}
winner = opt_results %>% top_n(1, -jitter(score))
winners = opt_results %>% group_by(run) %>% top_n(1, -jitter(score)) %>% ungroup
```
```{r, echo = FALSE}
winner = example_winner
winners = example_winners
```

```{r, fig.width = 9, fig.align="center"}
winner$score[[1]]
plot_graph(winner$edges[[1]])
```
<br>



### Parallelization

Depending on the input parameters, `optimize_admixturegraph` can take a long time to run. With the parameters above, 420,000 admixture graphs (200 * 100 * (20+1)) will be generated and evaluated. Thanks to the function [`future_map`](https://www.rdocumentation.org/packages/furrr/versions/0.1.0/topics/future_map) from the [`furrr`](https://github.com/DavisVaughan/furrr) package, this can be sped up by parallelizing across the independent repeats. The function [`future::plan`](https://www.rdocumentation.org/packages/future/versions/1.15.1/topics/plan) can be called to specify the details of the parallelization. This can be used to parallelize across CPU cores or across nodes on a compute cluster.

Setting it up like this will make it run multithreaded:
```{r, eval = FALSE}
future::plan('multiprocess')
```
<br>

On the O2 cluster, the following command will set up parallelization across compute nodes.

```{r, eval = FALSE}
future::plan(tweak(batchtools_slurm, workers = 50,
                   resources=list(ncpus = 1, memory = 1024,
                                  walltime = 10*60*60, partition = 'short')))
```
This specifies that up to 50 jobs should be run at a time, with each one requesting one CPU, 1024 MB of memory, and 10 hours on the `short` partition.

This requires the R package `future.batchtools` and a batchtools template file in the working directory, such as this one: `/n/groups/reich/robert/projects/admixprograms/batchtools.slurm.tmpl`.

After that, the `optimize_admixturegraph` function will submit each of the 200 repeats as a separate job.

As it will still take a while for this to finish, it is a good idea to submit this as one job which calls an R script. That R script will in turn spawn 200 new jobs and wait for them to finish and return their results.

First, the input should be saved to a file.
```{r, eval = FALSE}
save(pops, f3_jest, ppinv, file = 'qpgraph_input.RData')
```

The R script could then look like this.
```{r, eval = FALSE}
library(admixtools)
library(future.batchtools)

load('qpgraph_input.RData')

future::plan(tweak(batchtools_slurm, workers=50,
                   resources=list(ncpus = 1, memory=1024,
                                  walltime=10*60*60, partition='short')))

opt_results = optimize_admixturegraph(pops, outpop = pops[1],
                                      f3_jest, ppinv, numrep = 200,
                                      numgen = 20, numgraphs = 100,
                                      numadmix = 3, verbose = FALSE)

save(opt_results, file='opt_results.RData')
```

It could be in a file called `opt_graphs.Rscript` and be run like this, or submitted as a job.
```{r, eval = FALSE}
Rscript opt_graphs.Rscript
```

It takes more time to evaluate larger graphs, and in particular graphs with more admixture nodes. It's probably a good idea to start with a small number of repeats, generations, and graphs per generation to get a sense of the runtime before scaling it up. 

<br>



## Summarizing graphs

<br>

> *All models are wrong, but some are useful.*   --- George Box

<br>

> *All good models are alike; each bad model is bad in its own way.*   --- Leo Tolstoy

<br>

For any set of populations there may be many graphs which can explain the data about equally well. Rather than declaring one of these models as the correct model, it may be more useful to find features that are shared between all models with good fits. The following function attempts to find such features by summarizing a set of graphs. It counts how often each population triple occurs in which configuration across all graphs.

```{r, eval = FALSE}
triples = summarize_triples(winners)
```
```{r, echo = FALSE}
triples = example_triples
```

```{r}
triples %>% arrange(-clade) %>% slice(1)
```

The output for this population triple (1: `Rus`, 2: `Swi`, 3: `Den`) can be read like this:

* `numgraphs`: 200 graphs were compared
* `clade`: In 34.5% of all graphs `Rus` and `Swi` form a clade with respect to `Den`
* `x13`: In 0.5% of all graphs `Rus` is closer to `Den` than `Swi` is to `Den`, with ancestors of `Rus` admixing into `Den` 
* `x31`: In 50.5% of all graphs `Rus` is closer to `Den` than `Swi` is to `Den`, with ancestors of `Den` admixing into `Rus`
* `x23`: In 24.5% of all graphs `Swi` is closer to `Den` than `Rus` is to `Den`, with ancestors of `Swi` admixing into `Den` 
* `x32`: In 38.0% of all graphs `Swi` is closer to `Den` than `Rus` is to `Den`, with ancestors of `Den` admixing into `Swi`
* `toptopo`: The most common topology for this triple across all graphs. `011111` specifies a topology where condition `x13` is not satisfied, but conditions `x23`, `x31`, `x32`, `x12`, and `x21` are satisfied. A topology starting with `0000` is one where the first two populations form a clade.
* `toptopocnt`: The number of times `toptopo` was observed
* `topos`: The number of times each topology was observed


<br>



## Finding valid qpAdm models

Every admixture graph maps to a specific set of valid qpAdm models. The function `qpadm_models` lists all valid qpAdm models for a graph. Models which are contained within larger valid models are not shown. The number of valid qpAdm models can be very large for big graphs, so you should only run this on small graphs.

```{r, fig.width = 9, fig.align="center"}
mygraph = winners %>% slice(6) %$% igraph %>% pluck(1)
plot_graph(mygraph)
qpadm_models(mygraph, add_outgroup = TRUE, nested = FALSE, abbr = 3)
```

The only qpAdm models which are valid under this graph have `Den` as target and `Chi`, `Rus`, and `Swi` as right populations.


Valid qpAdm models need to satisfy the following criteria:

* There have to be more right populations than left populations
* The left populations should not form a clade with respect to any right population. This means that a qpgraph outgroup is not an informative right population, but it can be included anyway, with `add_outgroup`.
* Each set of left populations which forms a source for the target population has to form a clade with target at the exclusion of each right population. (Note: Check if that's correct and if it can be stated in a better way.)






