<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Inferring demographic history from genetic data • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="extra.js"></script><meta property="og:title" content="Inferring demographic history from genetic data">
<meta property="og:description" content="admixtools is a set of methods used to infer demographic history from genetic data using f-statistics.
  It was originally implemented as a command line program in C. This is a new implementation of the programs
  qpDstat, qpAdm, qpGraph, and contains several new features.
  It is efficient because it precomputes f2-statistics which can be re-used for many analyses.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="articles/io.html">Genotype data</a>
    </li>
    <li>
      <a href="articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="articles/resampling.html">Standard errors</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="admixtools-2" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#admixtools-2" class="anchor"></a>ADMIXTOOLS 2</h1></div>
<p>A new, lightning fast implementation of <a href="https://github.com/DReichLab/AdmixTools">ADMIXTOOLS</a>.</p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>ADMIXTOOLS is a collection of programs which use genetic data to infer how populations are related to one another. It has been used in countless publications to test whether populations form clades (<em>qpDstat</em>, <em>qpWave</em>), to estimate ancestry proportions (<em>qpAdm</em>), and to fit admixture graphs (<em>qpGraph</em>).</p>
<p>ADMIXTOOLS 2 provides the same functionality in a new look, and it’s orders of magnitude faster. This is achieved through separating the computation of f2-statistics from all other computations, and through a number of other optimizations. In the example below, rendering the plot takes much longer than computing the fit of a new <em>qpGraph</em> model:</p>
<p><img src="reference/figures/shinyapp1.gif" alt="app demo"></p>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<ul>
<li>Much faster than the original ADMIXTOOLS software</li>
<li>Simple R command line interface</li>
<li>Even simpler point-and-click interface</li>
<li>Several new features and methodological innovations that make it easier to find robust models:
<ul>
<li>Automated and semi-automated admixture graph inference</li>
<li>Simultaneous exploration of hundreds of <em>qpAdm</em> models</li>
<li>Unbiased comparison of any two <em>qpGraph</em> models using out-of-sample scores</li>
<li>Jackknife and bootstrap standard errors and confidence intervals for any <em>qpAdm</em>, <em>qpWave</em>, and <em>qpGraph</em> parameters</li>
</ul>
</li>
<li>Full support for genotype data in (PACKED)ANCESTRYMAP/EIGENSTRAT format and PLINK format</li>
<li>Wrapper functions around the original ADMIXTOOLS software (see also <a href="https://bodkan.net/admixr/index.html">admixr</a>) <!-- * Simple interface with [msprime](https://msprime.readthedocs.io/en/stable/index.html) for simulating under a given admixture graph -->
</li>
<li><a href="https://uqrmaie1.github.io/admixtools/articles/admixtools.html">Extensive documentation</a></li>
<li>New features available <a href="mailto:rmaier@broadinstitute.org">on request</a>!</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install and load ADMIXTOOLS 2, start R (version 3.5 or higher) and follow these steps:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>) <span class="co"># if "devtools" is not installed already</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"uqrmaie1/admixtools"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"admixtools"</span>)</a></code></pre></div>
<p>The above commands will install all R package dependencies which are required to run ADMIXTOOLS 2 on the command line. For the interactive app, additional packages are required, which can be installed like this:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"uqrmaie1/admixtools"</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>If you encounter any problems during the installation, this is most likely because some of the required R packages cannot be installed. If that happens, try manually re-installing some of the larger packages, and pay attention to any error message:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"Rcpp"</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"tidyverse"</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"igraph"</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"plotly"</span>)</a></code></pre></div>
<p>If you get the following error on Linux <code>Error: package or namespace load failed for 'admixtools' in dyn.load(file, DLLpath = DLLpath, ...):</code> try adding the following two lines to the file <code>~/.R/Makevars</code> (and create the file first if it doesn’t exist)</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">#LAPACK_LIBS=-llapack</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">PKG_LIBS =<span class="st"> </span><span class="er">$</span>(LAPACK_LIBS)</a></code></pre></div>
<p>If the installation still fails, please <a href="mailto:rmaier@broadinstitute.org">contact me</a>.</p>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>The first step is to extract f2-statistics from genotype files. These f2-statistics will be written to disk so that the slow part of the computation does not have to be repeated.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">genotype_data =<span class="st"> "/my/geno/prefix"</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">f2_dir =<span class="st"> "/my/f2/directory/"</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw"><a href="reference/extract_f2.html">extract_f2</a></span>(genotype_data, f2_dir)</a></code></pre></div>
<p>After that, we can fit an admixture graph like this:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">f2_blocks =<span class="st"> </span><span class="kw"><a href="reference/f2_from_precomp.html">f2_from_precomp</a></span>(f2_dir)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">fit =<span class="st"> </span><span class="kw"><a href="reference/qpgraph.html">qpgraph</a></span>(f2_blocks, example_graph)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">fit<span class="op">$</span>score</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="kw"><a href="reference/plot_graph.html">plot_graph</a></span>(fit<span class="op">$</span>edges)</a></code></pre></div>
<p><img src="reference/figures/graph1.png" alt="example graph"></p>
<p>Clearly not a historically accurate model, but it gets the idea across.</p>
<p><br></p>
<p>We can also use the f2-statistics to estimate admixture weights:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">left =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Altai_Neanderthal.DG"</span>, <span class="st">"Vindija.DG"</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">right =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Chimp.REF"</span>, <span class="st">"Mbuti.DG"</span>, <span class="st">"Russia_Ust_Ishim.DG"</span>, <span class="st">"Switzerland_Bichon.SG"</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">target =<span class="st"> "Denisova.DG"</span></a></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="reference/qpadm.html">qpadm</a></span>(f2_blocks, left, right, target)<span class="op">$</span>weights</a></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">#&gt; # A tibble: 2 x 5</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co">#&gt;   target      left                 weight    se     z</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">#&gt;   &lt;chr&gt;       &lt;chr&gt;                 &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; 1 Denisova.DG Altai_Neanderthal.DG   49.6  23.3  2.13</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">#&gt; 2 Denisova.DG Vindija.DG            -48.6  23.3 -2.08</span></a></code></pre></div>
<p><br></p>
<p>Or we can use them to get f4-statistics:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="reference/qpdstat.html">f4</a></span>(f2_blocks)</a></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co">#&gt; ℹ Getting population combinations...</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#&gt; ℹ 105 population combinations found</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">#&gt; ℹ Loading f2 data for 42 population pairs...</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">#&gt; ℹ Computing f4-statistics</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#&gt; # A tibble: 105 x 8</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">#&gt;    pop1        pop2         pop3    pop4            est      se      z         p</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">#&gt;  1 Altai_Nean… Chimp.REF    Deniso… Mbuti.DG     0.0196 6.07e-4   32.4 1.32e-229</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co">#&gt;  2 Altai_Nean… Denisova.DG  Chimp.… Mbuti.DG    -0.0129 3.64e-4  -35.6 2.22e-277</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10"><span class="co">#&gt;  3 Altai_Nean… Mbuti.DG     Chimp.… Denisova.DG -0.0326 5.22e-4  -62.5 0.       </span></a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="co">#&gt;  4 Altai_Nean… Chimp.REF    Deniso… Russia_Ust…  0.0180 6.87e-4   26.3 6.43e-152</span></a>
<a class="sourceLine" id="cb11-12" data-line-number="12"><span class="co">#&gt;  5 Altai_Nean… Denisova.DG  Chimp.… Russia_Ust… -0.0152 4.46e-4  -34.0 4.67e-254</span></a>
<a class="sourceLine" id="cb11-13" data-line-number="13"><span class="co">#&gt;  6 Altai_Nean… Russia_Ust_… Chimp.… Denisova.DG -0.0332 5.55e-4  -60.0 0.       </span></a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="co">#&gt;  7 Altai_Nean… Chimp.REF    Deniso… Switzerlan…  0.0181 6.63e-4   27.3 1.09e-164</span></a>
<a class="sourceLine" id="cb11-15" data-line-number="15"><span class="co">#&gt;  8 Altai_Nean… Denisova.DG  Chimp.… Switzerlan… -0.0150 4.64e-4  -32.3 6.06e-229</span></a>
<a class="sourceLine" id="cb11-16" data-line-number="16"><span class="co">#&gt;  9 Altai_Nean… Switzerland… Chimp.… Denisova.DG -0.0331 5.74e-4  -57.7 0.       </span></a>
<a class="sourceLine" id="cb11-17" data-line-number="17"><span class="co">#&gt; 10 Altai_Nean… Chimp.REF    Deniso… Vindija.DG  -0.0771 6.98e-4 -110.  0.       </span></a>
<a class="sourceLine" id="cb11-18" data-line-number="18"><span class="co">#&gt; # … with 95 more rows</span></a></code></pre></div>
</div>
<div id="interactive-browser-app" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-browser-app" class="anchor"></a>Interactive browser app</h2>
<p>ADMIXTOOLS 2 also has a simple point-and-click interface. This makes it easy to explore many <em>qpAdm</em> or <em>qpGraph</em> models at the same time, for example by allowing you to build and change admixture graphs interactively. Typing the following command in the R console launches the app:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="reference/run_shiny_admixtools.html">run_shiny_admixtools</a></span>()</a></code></pre></div>
<p><img src="reference/figures/shinyapp2.gif" alt="app demo"></p>
</div>
<div id="documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation" class="anchor"></a>Documentation</h2>
<p>One of the design goals behind ADMIXTOOLS 2 is to make the algorithms more transparent, so that the steps leading from from genotype data to conclusions about demographic history are easier to follow.</p>
<p>To this end, all ADMIXTOOLS 2 functions are <a href="https://uqrmaie1.github.io/admixtools/reference/index.html">documented</a>. You can also take a look at the <a href="articles/admixtools.html">tutorial</a>, read more about how ADMIXTOOLS 2 computes <a href="https://uqrmaie1.github.io/admixtools/articles/fstats.html">f-statistics</a> and <a href="https://uqrmaie1.github.io/admixtools/articles/resampling.html">standard errors</a>, and what you can do with <a href="https://uqrmaie1.github.io/admixtools/articles/graphs.html">admixture graphs</a>.</p>
<p>For even greater transparency, many of the core functions are implemented twice: In C++ for performance (used by default), and in R, which makes it easier to trace the computations step by step. For example, if you want to know how weights are computed in <code><a href="reference/qpadm.html">qpadm()</a></code>, you can type <code>qpadm</code> in R to get the function code and you will see another function which is called <code>qpadm_weights()</code>. By default, this function will be replaced by its faster C++ version <code>cpp_qpadm_weights()</code>. But you can still see what it’s doing without reading the C++ code by typing <code>admixtools:::qpadm_weights</code>. And you can tell <code><a href="reference/qpadm.html">qpadm()</a></code> to use the R versions instead of the C++ versions by calling <code><a href="reference/qpadm.html">qpadm(cpp = FALSE)</a></code>.</p>
</div>
<div id="contact" class="section level2">
<h2 class="hasAnchor">
<a href="#contact" class="anchor"></a>Contact</h2>
<p>For questions, feature requests, and bug reports, please contact Robert Maier under <a href="mailto:rmaier@broadinstitute.org" class="email">rmaier@broadinstitute.org</a>.</p>
</div>
<div id="see-also" class="section level2">
<h2 class="hasAnchor">
<a href="#see-also" class="anchor"></a>See also</h2>
<ul>
<li>
<a href="https://github.com/DReichLab/AdmixTools">ADMIXTOOLS</a> The original ADMIXTOOLS software</li>
<li>
<a href="https://bodkan.net/admixr/index.html">admixr</a> An R package with ADMIXTOOLS wrapper functions and many useful tutorials</li>
<li>
<a href="https://github.com/mailund/admixture_graph">admixturegraph</a> An R package for automatic graph inference</li>
<li>
<a href="https://github.com/juliayyan/PhylogeneticTrees.jl">miqoGraph</a> A Julia package for automatic graph inference</li>
<li>
<a href="http://cb.csail.mit.edu/cb/mixmapper/">MixMapper</a> Another method to infer admixture graphs</li>
<li>
<a href="https://bitbucket.org/nygcresearch/treemix/wiki/Home">TreeMix</a> Another method to infer admixture graphs</li>
<li>
<a href="http://content.csbs.utah.edu/~rogers/src/legofit/index.html">Legofit</a> A program to estimate the history of population size, subdivision, and gene flow</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Robert Maier <br><small class="roles"> Author, maintainer </small>  </li>
<li>Nick Patterson <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
