<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing *f*-statistics • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="Computing *f*-statistics">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fstats.html">Computing *f*-statistics</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Computing <em>f</em>-statistics</h1>
            
      
      
      <div class="hidden name"><code>fstats.Rmd</code></div>

    </div>

    
    
<p><em>f</em>-statistics are the foundation of ADMIXTOOLS. In ADMIXTOOLS 2, f2-statistics are of particular importance as f3- and f4-statistics are computed from f2-statistics. f2-statistics are conceptually very simple, but estimating them without bias can be complicated. Here, we describe what these complications are, and how we deal with them.</p>
<div id="estimating-f2" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-f2" class="anchor"></a>Estimating <em>f<sub>2</sub></em>
</h2>
<p><span class="math inline">\(f_2\)</span> is the expected squared difference in allele frequencies between two populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. It is the between-population variance, or the numerator in the more widely used <span class="math inline">\(F_{ST}\)</span>, which is the ratio of between-population variance to total variance: <span class="math inline">\(F_{ST}(A, B) = \frac{f_2(A,B)}{f_2(A,B) + \frac{1}{n} \sum_{j=1}^n a_j(1-a_j) + b_j (1-b_j)}\)</span>.</p>
<p>In the simplest case, we can estimate <span class="math inline">\(f_2\)</span> across <span class="math inline">\(n\)</span> SNPs like this: <span class="math display">\[f_2(A,B) = \frac{1}{n} \sum_{j=1}^n(a_{j} - b_{j})^2 = \frac{1}{n} \sum_{j=1}^n a^2_{j} - 2 a_j b_j + b^2_{j}\]</span></p>
<p>The simplest case is one where:</p>
<ol style="list-style-type: decimal">
<li>The infinite sites model is approximately correct</li>
<li>We have very accurate estimates of allele frequencies</li>
<li>There is no missing data</li>
</ol>
<p>Here we will look at what happens if 2. and 3. are not true.</p>
</div>
<div id="bias-due-to-inaccurate-allele-frequency-estimates" class="section level2">
<h2 class="hasAnchor">
<a href="#bias-due-to-inaccurate-allele-frequency-estimates" class="anchor"></a>Bias due to inaccurate allele frequency estimates</h2>
<p>Inaccurate allele frequency estimates lead to over-esimation of <span class="math inline">\(f_2\)</span>: If two populations have diverged very recently, their allele frequencies will be very similar. But if the allele frequencies of one or both population are estimated very inaccurately, the allele frequency differences (and <span class="math inline">\(f_2\)</span>) will appear larger than they really are.</p>
<p>The most common reason for inaccurate allele frequency estimates is low sample count (or a low number of independently sampled alleles). The upward bias due to low sample count at a single SNP in one population is <span class="math inline">\(\frac{p(1-p)}{n-1}\)</span>, where <span class="math inline">\(p\)</span> is the allele frequency and <span class="math inline">\(n\)</span> is the number of independently sampled alleles. Therefore, an unbiased estimate of <span class="math inline">\(f_2\)</span> at a single SNP is given by</p>
<p><span class="math display">\[f_2(A, B) = (a-b)^2 - \frac{a(1-a)}{n_A-1} - \frac{b(1-b)}{n_B-1}\]</span> In most cases, <span class="math inline">\(n\)</span>, the number of independently sampled alleles, will be twice the number of samples, with two exceptions:</p>
<div id="pseudohaploid-data" class="section level3">
<h3 class="hasAnchor">
<a href="#pseudohaploid-data" class="anchor"></a>Pseudohaploid data</h3>
<p>Many ancient DNA data sets don’t call heterozygous genotypes in order to avoid a number of biases. If two different alleles are observed in one sample at a particular site, one of the two alleles is chosen at random and the genotype is declared homozygous for that allele. In pseudohaploid samples, the number of independently sampled alleles is therefore equal to the number of samples.</p>
</div>
<div id="inbreeding" class="section level3">
<h3 class="hasAnchor">
<a href="#inbreeding" class="anchor"></a>Inbreeding</h3>
<p>If an individuals has closely related parents, his or her two parental chromosomes are more similar to each other than two chromosomes sampled at random from the population. In a population in which inbreeding is common, <span class="math inline">\(n\)</span> can therefore be less than two.</p>
<p>Some of the programs in the original ADMIXTOOLS have an <code>inbreed</code> option which uses an estimator of <span class="math inline">\(f_2\)</span> that is robust to inbreeding. This option is also recommended to use for pseudohaploid data. ADMIXTOOLS 2 automatically detects pseudohaploid samples by checking if any heterozyous genotypes are observed among the first 1000 SNPs. If not, the sample is assumed to be pseudohaploid.</p>
</div>
</div>
<div id="bias-due-to-missing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#bias-due-to-missing-data" class="anchor"></a>Bias due to missing data</h2>
<p>Missing data alone will not create any bias in f2 estimates, as long as the data are missing at random. However, if rare or common SNPs are more likely to be missing, f2 can be over- or underestimated. In contrast to <span class="math inline">\(F_{ST}\)</span>, which is the same in expectation for rare and for common SNPs, <span class="math inline">\(f_2\)</span> tends to be larger for common SNPs. Strictly speaking, there is no such thing as an unbiasd <span class="math inline">\(f_2\)</span> estimate, since it always depends on the allele frequency spectrum of ascertained SNPs. What’s important in practice is that there are no systematic differences in ascertainment among the populations in an analysis. ADMIXTOOLS draws most of its inferences from comparing allele frequency differences between populations, which means that biases that affect all populations equally are not very important. If, however, one pair of populations has a high f2 because most rare alleles are missing in one of the populations, and another pair of populations has a lower f2 beause those rare alleles are not missing, bias can be expected.</p>
<p>One way to prevent this bias is to not use any SNPs which are missing in one or more populations. But this may not leave enough SNPs to have sufficient statistical power. A middle ground can be reached by excluding only SNPs with very high missingness, and thus high potential for bias.</p>
</div>
<div id="estimating-f4" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-f4" class="anchor"></a>Estimating <em>f<sub>4</sub></em>
</h2>
<p><span class="math inline">\(f_4\)</span> is the covariance of allele frequency differences between two pairs of populations. It can also be expressed as a sum of <span class="math inline">\(f_2\)</span>-statistics:</p>
<p><span class="math display">\[
\begin{equation}
\begin{aligned}
f_4(A, B; C, D) &amp;= \frac{1}{n}\sum_{j=1}^n(a_{j} - b_{j})(c_{j} - d_{j}) \\ &amp;= \frac{1}{n} \sum_{j=1}^n(a_j c_j + b_j d_j - a_j d_j - b_j c_j) \\ &amp;= \frac{1}{2n} (\sum_{j=1}^n(a_j - d_j)^2 + \sum_{j=1}^n(b_j - c_j)^2 - \sum_{j=1}^n(a_j - c_j)^2 - \sum_{j=1}^n(b_j - d_j)^2) \\ &amp;= \frac{1}{2}(f_2(A, D) + f_2(B, C) - f_2(A, C) - f_2(B, D) )
\label{eq:f42}
\end{aligned}
\end{equation}
\]</span></p>
<p>This makes it clear that <span class="math inline">\(f_4\)</span> is a generalization of <span class="math inline">\(f_2\)</span>: <span class="math inline">\(f_4(A, B; A, B) = f_2(A, B)\)</span></p>
<p>D-statistics differ from <span class="math inline">\(f_4\)</span> only by a scaling factor: <span class="math inline">\(D(A, B; C, D) = \frac{f_4(A, B; C, D)}{\frac{1}{n} \sum_{j=1}^n (a_j + b_j - 2 a_j b_j) (c_j + d_j - 2 c_j d_j)}\)</span></p>
<p>Naively one might think that estimating <span class="math inline">\(f_4\)</span> without bias is at least as difficult as estimating <span class="math inline">\(f_2\)</span> without bias, but that is not the case. Since each population is represented through two <span class="math inline">\(f_2\)</span> terms of opposite sign in <span class="math inline">\(f_4\)</span>, many biases that are specific to one population will cancel out. This also applies to the low allele count correction factor <span class="math inline">\(\frac{p(1-p)}{n-1}\)</span>, which is not necessary when estimating <span class="math inline">\(f_4\)</span>.</p>
<p>For a visual interpretation of this effect, consider the following phylogeny:</p>
<pre><code>A         B
 \_______/
 /       \
C         D</code></pre>
<p><span class="math inline">\(f_2\)</span> is the total branch length between any two populations, while <span class="math inline">\(f_4(A, B; C, D)\)</span> is the length of the horizontal line. Any increase or decrease in the lengths of the terminal branches leading to <span class="math inline">\(A\)</span>, <span class="math inline">\(B\)</span>, <span class="math inline">\(C\)</span>, or <span class="math inline">\(D\)</span> will affect two <span class="math inline">\(f_2\)</span> statistics equally, but will leave <span class="math inline">\(f_4(A, B; C, D)\)</span> unaffected. This robustness is a major strength of methods based on f4-statistics (and to a large extent also of f3-based methods): Terminal branch lengths are affected by many factors such as divergence time, changes in population size, as well as by data quality, but none of this affects <span class="math inline">\(f_4\)</span>, the covariance of allele freuquency differences.</p>
</div>
<div id="more" class="section level2">
<h2 class="hasAnchor">
<a href="#more" class="anchor"></a>more</h2>
<p>If you’re not on O2, or want use different samples or populations, you first have to compute f2 statistics from genotype data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">my_f2_dir =<span class="st"> 'store/f2data/here/'</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">pops =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Denisova.DG'</span>,</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">         <span class="st">'Altai_Neanderthal.DG'</span>,</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">         <span class="st">'Vindija.DG'</span>,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">         <span class="st">'Chimp.REF'</span>,</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">         <span class="st">'Mbuti.DG'</span>,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">         <span class="st">'Russia_Ust_Ishim.DG'</span>,</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">         <span class="st">'Switzerland_Bichon.SG'</span>)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="kw"><a href="../reference/extract_f2.html">extract_f2</a></span>(prefix, <span class="dt">outdir =</span> my_f2_dir, <span class="dt">pops =</span> pops)</a></code></pre></div>
<p>This will look for genotype files in <code>packedancestrymap</code> or <code>PLINK</code> format, compute allele frequencies and blocked f2 statistics for all pairs of populations defined in the <code>.ind</code> or <code>.fam</code> file, and write them to <code>outdir</code>.</p>
<p><br></p>
<p>With <code>f2_from_geno</code>, you can skip writing data to disk, and read f2 blocks from genotype files directly. This can be useful if the number of populations is not too large:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">f2_blocks =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_geno.html">f2_from_geno</a></span>(prefix, <span class="dt">pops =</span> pops)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">qpg_results =<span class="st"> </span><span class="kw"><a href="../reference/qpgraph.html">qpgraph</a></span>(f2_blocks, example_graph)</a></code></pre></div>
<p><br></p>
<div id="arbitrary-populations" class="section level3">
<h3 class="hasAnchor">
<a href="#arbitrary-populations" class="anchor"></a>Arbitrary populations</h3>
<p>What if you decide that a particular sample should be added or removed from a population? With the approach described so far, you would have to go back to the genotype data to get new estimates of f2 for all pairs of the modified population and any other populations. Even if you had computed f2 for all pairs of individuals (rather than for pairs of populations), it would not be possible to combine the f2 statistics from individual pairs into f2 statistics on population pairs formed from these individuals. It <em>is</em> possible, however, to store <em>other</em> statistics for all individuals and individual pairs, which can be combined into estimates of f2 for any populations formed from these individuals. These statistics are simply the mean alternative allele counts (<code>a</code>) and total allele counts (<code>n</code>) for each individual and block, and the mean products of allele counts for all individual pairs (<code>aa</code> and <code>nn</code>).</p>
<p>The upside of this approach is that it allows you to change population definitions however you like and still get very fast results without accessing the genotype data more than once. The downside is that it makes it necessary to store data for all individual pairs, rather than just all population pairs.</p>
<p>The following function extracts data for a set of individuals which can then be combined to get estimates of f2 for any populations:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">inds =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp.REF'</span>, <span class="st">'Altai_snpAD.DG'</span>, <span class="st">'Vindija_snpAD.DG'</span>,</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">         <span class="st">'S_Mbuti-3.DG'</span>, <span class="st">'B_Mbuti-4.DG'</span>, <span class="st">'S_Mbuti-2.DG'</span>, <span class="st">'S_Mbuti-1.DG'</span>,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">         <span class="st">'Denisova_snpAD.DG'</span>, <span class="st">'UstIshim_snpAD.DG'</span>, <span class="st">'Bichon.SG'</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">my_counts_dir =<span class="st"> 'my/counts/dir'</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw"><a href="../reference/extract_counts.html">extract_counts</a></span>(prefix, my_counts_dir, inds)</a></code></pre></div>
<p>You can again use <code>f2_from_precomp</code> to load a subset of the extracted data into memory. If the directory you provide contains allele count data rather than f2 statistics, this function will not only read the data, but combine it into f2 statistics for populations that you define. The populations can be defined by providing a vector of individuals and another vector of equal length with population labels.</p>
<p>For example, the following code should return f2 statistics very similar to the f2 statistics in the earlier examples. They will not be exactly identical, because some approximations are necessary when computing f2 statistics this way.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">pops =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp'</span>, <span class="st">'Altai'</span>, <span class="st">'Vindija'</span>,</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">         <span class="st">'Mbuti'</span>, <span class="st">'Mbuti'</span>, <span class="st">'Mbuti'</span>, <span class="st">'Mbuti'</span>,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">         <span class="st">'Denisova'</span>, <span class="st">'Russia'</span>, <span class="st">'Switzerland'</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">f2_blocks_v2 =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_precomp.html">f2_from_precomp</a></span>(my_counts_dir, <span class="dt">inds =</span> inds, <span class="dt">pops =</span> pops)</a></code></pre></div>
<p>It’s now easy to test how the inclusion or exclusion of specific samples affects the fit of a model. The following example uses only two Mbuti samples, and combines the two Neanderthal samples into one group.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">inds_v3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp.REF'</span>, <span class="st">'Altai_snpAD.DG'</span>, <span class="st">'Vindija_snpAD.DG'</span>,</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">            <span class="st">'S_Mbuti-2.DG'</span>, <span class="st">'S_Mbuti-1.DG'</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">            <span class="st">'Denisova_snpAD.DG'</span>, <span class="st">'UstIshim_snpAD.DG'</span>, <span class="st">'Bichon.SG'</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">pops_v3 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp'</span>, <span class="st">'Neanderthal'</span>, <span class="st">'Neanderthal'</span>,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">            <span class="st">'Mbuti'</span>, <span class="st">'Mbuti'</span>,</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">            <span class="st">'Denisova'</span>, <span class="st">'Russia'</span>, <span class="st">'Switzerland'</span>)</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">f2_blocks_v3 =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_precomp.html">f2_from_precomp</a></span>(my_counts_dir, <span class="dt">inds =</span> inds_v3, <span class="dt">pops =</span> pops_v3)</a></code></pre></div>
<p><br></p>
<p>This approach is very flexible, but it gets slow when some populations consist of large numbers of samples. If you find yourself in this situation, you can group individuals into populations and store the combined allele count data for those populations on disk. The following example stores a <code>Mbuti</code> group on disk, and then uses that group, rather than each sample in it, to compute f2 statistics.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">mbuti =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'S_Mbuti-3.DG'</span>, <span class="st">'B_Mbuti-4.DG'</span>, <span class="st">'S_Mbuti-2.DG'</span>, <span class="st">'S_Mbuti-1.DG'</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw"><a href="../reference/group_samples.html">group_samples</a></span>(my_counts_dir, mbuti, <span class="st">'Mbuti'</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4">inds_v4 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Chimp.REF'</span>, <span class="st">'Altai_snpAD.DG'</span>, <span class="st">'Vindija_snpAD.DG'</span>, <span class="st">'Mbuti'</span>,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">            <span class="st">'Denisova_snpAD.DG'</span>, <span class="st">'UstIshim_snpAD.DG'</span>, <span class="st">'Bichon.SG'</span>)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">f2_blocks_v4 =<span class="st"> </span><span class="kw"><a href="../reference/f2_from_precomp.html">f2_from_precomp</a></span>(my_counts_dir, <span class="dt">inds =</span> inds_v4)</a></code></pre></div>
<p>Data from these groups can be deleted like this:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="../reference/delete_groups.html">delete_groups</a></span>(my_counts_dir, <span class="st">'Mbuti'</span>)</a></code></pre></div>
<p><br></p>
<p>It makes it possible to use the exact same set of SNPs for each population quadruple. This is not guaranteed when f4-statistics are computed from f2-statistics. In the presence of large amounts of missing data, this can lead to biased f4-statistics. Computing f4-statistics from genotype data directly avoids this issue. There are other ways to avoid any potential bias without sacrificing speed.</p>
<p>One is to to not use SNPs which are missing in any populations:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="../reference/extract_f2.html">extract_f2</a></span>(prefix, my_f2_dir, <span class="dt">maxmiss =</span> <span class="dv">0</span>)</a></code></pre></div>
<p>If too many SNPs are lost this way, it is also possible to load</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#estimating-f2">Estimating <em>f<sub>2</sub></em></a></li>
      <li><a href="#bias-due-to-inaccurate-allele-frequency-estimates">Bias due to inaccurate allele frequency estimates</a></li>
      <li><a href="#bias-due-to-missing-data">Bias due to missing data</a></li>
      <li><a href="#estimating-f4">Estimating <em>f<sub>4</sub></em></a></li>
      <li><a href="#more">more</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
