<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genotype data • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="Genotype data">
<meta property="og:description" content="admixtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Genotype data</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
    <li>
      <a href="../articles/simulations.html">Simulations</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/uqrmaie1/admixtools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="io_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="io_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="io_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genotype data</h1>
                        <h4 class="author">Robert Maier</h4>
            
            <h4 class="date">2020-11-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/uqrmaie1/admixtools/blob/master/vignettes/io.Rmd"><code>vignettes/io.Rmd</code></a></small>
      <div class="hidden name"><code>io.Rmd</code></div>

    </div>

    
    
<p><strong>Under construction</strong></p>
<p><em>ADMIXTOOLS 2</em> can read genotype data in binary <em>PLINK</em> format which is described <a href="https://www.cog-genomics.org/plink2/input">here</a>, and in <em>EIGENSTRAT</em>/<em>PACKEDANCESTRYMAP</em> formats, described <a href="https://reich.hms.harvard.edu/software/InputFileFormats">here</a>.</p>
<p>While there are other R packages for reading and writing binary <em>PLINK</em> files (<a href="https://github.com/gabraham/plink2R">plink2R</a>, <a href="https://github.com/OchoaLab/genio">genio</a>), no such packages exist to date for <em>PACKEDANCESTRYMAP</em> format files, so some of the functions shown here may be useful outside of the <em>ADMIXTOOLS</em> framework.</p>
<p>Note that the functions below impose the restriction than individual IDs may not be duplicated across populations or families.</p>
<p><em>ADMIXTOOLS 2</em> currently doesn’t handle many common genotype data wrangling tasks such merging different data sets. That’s partly because <em>PLINK</em> already excels at this task, and partly because it requires a large amount of effort to handle all possible scenarios that can come up, without introducing potentially dangerous bugs. If converting your data to <em>PLINK</em> format, and using <em>PLINK</em> for these tasks is not a practical solution in your case, please let me know about it.</p>
<div id="reading-genotype-files" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-genotype-files" class="anchor"></a>Reading genotype files</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geno</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_plink.html">read_plink</a></span><span class="op">(</span><span class="st">"plink/prefix"</span><span class="op">)</span>
<span class="va">geno</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_eigenstrat.html">read_eigenstrat</a></span><span class="op">(</span><span class="st">"eigenstrat/prefix"</span><span class="op">)</span>
<span class="va">geno</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_packedancestrymap.html">read_packedancestrymap</a></span><span class="op">(</span><span class="st">"packedancestrymap/prefix"</span><span class="op">)</span>

<span class="co"># read SNPs 1000 to 2000</span>
<span class="va">geno</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_packedancestrymap.html">read_packedancestrymap</a></span><span class="op">(</span><span class="st">"packedancestrymap/prefix"</span>, first <span class="op">=</span> <span class="fl">1000</span>, last <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span>
<span class="co"># read only ind1 and ind2</span>
<span class="va">geno</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_packedancestrymap.html">read_packedancestrymap</a></span><span class="op">(</span><span class="st">"packedancestrymap/prefix"</span>, inds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'ind1'</span>, <span class="st">'ind2'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="computing-allele-frequencies" class="section level3">
<h3 class="hasAnchor">
<a href="#computing-allele-frequencies" class="anchor"></a>Computing allele frequencies</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">afs</span> <span class="op">=</span> <span class="fu"><a href="../reference/plink_to_afs.html">plink_to_afs</a></span><span class="op">(</span><span class="st">"plink/prefix"</span><span class="op">)</span>
<span class="va">afs</span> <span class="op">=</span> <span class="fu"><a href="../reference/eigenstrat_to_afs.html">eigenstrat_to_afs</a></span><span class="op">(</span><span class="st">"eigenstrat/prefix"</span><span class="op">)</span>
<span class="va">afs</span> <span class="op">=</span> <span class="fu"><a href="../reference/packedancestrymap_to_afs.html">packedancestrymap_to_afs</a></span><span class="op">(</span><span class="st">"packedancestrymap/prefix"</span><span class="op">)</span></code></pre></div>
</div>
<div id="converting-packedancestrymap-to-plink" class="section level3">
<h3 class="hasAnchor">
<a href="#converting-packedancestrymap-to-plink" class="anchor"></a>Converting <em>PACKEDANCESTRYMAP</em> to <em>PLINK</em>
</h3>
<p>This function converts genotype files from <em>PACKEDANCESTRYMAP</em> or <em>EIGENSTRAT</em> to <em>PLINK</em> format. If <code>inds</code> or <code>pops</code> is specified, only those samples will be extracted.</p>
<p><em>ADMIXTOOLS 2</em> currently doesn’t provide the option to write binary <em>PACKEDANCESTRYMAP</em> files, so the conversion only goes one way.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/packedancestrymap_to_plink.html">eigenstrat_to_plink</a></span><span class="op">(</span><span class="st">"eigenstrat_input_prefix"</span>, <span class="st">"plink_output_prefix"</span><span class="op">)</span>
<span class="fu"><a href="../reference/packedancestrymap_to_plink.html">packedancestrymap_to_plink</a></span><span class="op">(</span><span class="st">"packedancestrymap_input_prefix"</span>, <span class="st">"plink_output_prefix"</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'c30b6be96312414d83a0ba52b94b558b',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
