<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>qpWave and qpAdm • admixtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><script src="../extra.js"></script><meta property="og:title" content="qpWave and qpAdm">
<meta property="og:description" content="admixtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/fstats.html">f-statistics</a>
    </li>
    <li>
      <a href="../articles/graphs.html">Admixture graphs</a>
    </li>
    <li>
      <a href="../articles/io.html">Genotype data</a>
    </li>
    <li>
      <a href="../articles/parallel.html">Parallelization</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
    <li>
      <a href="../articles/qpadm.html">qpWave and qpAdm</a>
    </li>
    <li>
      <a href="../articles/resampling.html">Standard errors</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/uqrmaie1/admixtools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="qpadm_files/header-attrs-2.5/header-attrs.js"></script><link href="qpadm_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="qpadm_files/anchor-sections-1.0/anchor-sections.js"></script><script src="qpadm_files/htmlwidgets-1.5.2/htmlwidgets.js"></script><script src="qpadm_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="qpadm_files/typedarray-0.1/typedarray.min.js"></script><script src="qpadm_files/jquery-1.11.3/jquery.min.js"></script><link href="qpadm_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="qpadm_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="qpadm_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="qpadm_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>qpWave and qpAdm</h1>
                        <h4 class="author">Robert Maier</h4>
            
            <h4 class="date">2021-01-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/uqrmaie1/admixtools/blob/master/vignettes/qpadm.Rmd"><code>vignettes/qpadm.Rmd</code></a></small>
      <div class="hidden name"><code>qpadm.Rmd</code></div>

    </div>

    
    
<p><strong>Under construction</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></code></pre></div>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✓ ggplot2 3.3.2     ✓ purrr   0.3.4
## ✓ tibble  3.0.4     ✓ dplyr   1.0.2
## ✓ tidyr   1.1.2     ✓ stringr 1.4.0
## ✓ readr   1.4.0     ✓ forcats 0.5.0</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## x tidyr::extract()   masks magrittr::extract()
## x dplyr::filter()    masks stats::filter()
## x dplyr::lag()       masks stats::lag()
## x purrr::set_names() masks magrittr::set_names()</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/uqrmaie1/admixtools">admixtools</a></span><span class="op">)</span></code></pre></div>
<div id="qpwave" class="section level2">
<h2 class="hasAnchor">
<a href="#qpwave" class="anchor"></a>qpWave</h2>
<div id="qpwave-and-f4" class="section level3">
<h3 class="hasAnchor">
<a href="#qpwave-and-f4" class="anchor"></a>qpWave and f4</h3>
<p><span class="math inline">\(f_4\)</span>-statistics can be used to test whether two pairs of populations form a clade with respect to one another: A significant deviation from zero is evidence against cladality. <em>qpWave</em> generalizes this test from two groups with two populations each (four in total) to two groups with any number of populations. In the special case of applying <em>qpWave</em> to only four populations, we get the same p-value that we would get for an <span class="math inline">\(f_4\)</span>-statistic of those four populations:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/qpdstat.html">f4</a></span><span class="op">(</span><span class="va">example_f2_blocks</span>, <span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span>, <span class="st">'Chimp.REF'</span>, <span class="st">'Mbuti.DG'</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 x 8
##   pop1                pop2       pop3      pop4          est      se     z     p
##   &lt;chr&gt;               &lt;chr&gt;      &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 Altai_Neanderthal.… Vindija.DG Chimp.REF Mbuti.DG 0.000124 1.35e-4 0.920 0.358</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/qpwave.html">qpwave</a></span><span class="op">(</span><span class="va">example_f2_blocks</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Altai_Neanderthal.DG'</span>, <span class="st">'Vindija.DG'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Chimp.REF'</span>, <span class="st">'Mbuti.DG'</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">rankdrop</span></code></pre></div>
<pre><code>## ℹ Computing f4 stats...
## ℹ Computing number of admixture waves...</code></pre>
<pre><code>## # A tibble: 1 x 7
##   f4rank   dof chisq     p dofdiff chisqdiff p_nested
##    &lt;int&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;int&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1      0     1 0.846 0.358      NA        NA       NA</code></pre>
</div>
<div id="estimating-the-number-of-independent-gene-flows" class="section level3">
<h3 class="hasAnchor">
<a href="#estimating-the-number-of-independent-gene-flows" class="anchor"></a>Estimating the number of independent gene flows</h3>
<p>Testing whether two pairs of populations form clades with each other is the same as asking whether there is only one gene flow event which separates the two pairs (<span class="math inline">\(f_4 = 0\)</span>) or whether there must have been more than a single gene flow event between the two pairs (<span class="math inline">\(f_4 != 0\)</span>).</p>
<p>In the graph on the left, <span class="math inline">\(L1\)</span> and <span class="math inline">\(L2\)</span> form a clade relative to <span class="math inline">\(R1\)</span> and <span class="math inline">\(R2\)</span>, which means <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> are only connected through one gene flow event. On the right they do not form a clade, and they are connected through two independent gene flow events (one going through the root, and the other going through the edge <code>n1b|n2b</code>).</p>
<div style="display: flex">
<div style="width: 50%">
<div id="htmlwidget-a53e43b1a2fcfcafd807" style="width:100%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-a53e43b1a2fcfcafd807">{"x":{"data":[{"x":[-1,-1.5,null,-1,-0.5,null,1,0.5,null,1,1.5],"y":[1,0,null,1,0,null,1,0,null,1,0],"text":["n1 -> L1","n1 -> L1",null,"n1 -> L2","n1 -> L2",null,"n2 -> R1","n2 -> R1",null,"n2 -> R2","n2 -> R2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"(normal,1)","legendgroup":"(normal,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,null,0,1],"y":[2,1,null,2,1],"text":["R -> n1","R -> n1",null,"R -> n2","R -> n2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"solid"},"hoveron":"points","name":"(normal,2)","legendgroup":"(normal,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,1],"y":[2,1,1],"text":["R","n1","n2"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":0,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,0.5,-1.25,-0.75,0.75,1.25],"y":[1.5,1.5,0.5,0.5,0.5,0.5],"text":["","","","","",""],"hovertext":["R -> n1","R -> n2","n1 -> L1","n1 -> L2","n2 -> R1","n2 -> R2"],"textfont":{"size":9.4488188976378,"color":"rgba(0,0,0,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,-0.5,0.5,1.5],"y":[-0.1,-0.1,-0.1,-0.1],"text":["L1","L2","R1","R2"],"hovertext":["","","",""],"textfont":{"size":9.4488188976378,"color":"rgba(248,118,109,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"1","legendgroup":"1","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.8246364188244,"r":7.30593607305936,"b":15.4776044553541,"l":10.958904109589},"paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.9,1.9],"tickmode":"array","ticktext":["-1","0","1"],"tickvals":[-1,0,1],"categoryorder":"array","categoryarray":["-1","0","1"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-0.205,2.105],"tickmode":"array","ticktext":["0.0","0.5","1.0","1.5","2.0"],"tickvals":[0,0.5,1,1.5,2],"categoryorder":"array","categoryarray":["0.0","0.5","1.0","1.5","2.0"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"4ce1b576dbe":{"linetype":{},"colour":{},"text":{},"x":{},"y":{},"xend":{},"yend":{},"from":{},"to":{},"type":"scatter"},"4ce118e5dac3":{"x":{},"y":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"4ce11a7d44fa":{"x":{},"y":{},"label":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"4ce1163b86ac":{"label":{},"colour":{},"from":{},"x":{},"y":{},"xend":{},"yend":{},"from.1":{},"to":{}}},"cur_data":"4ce1b576dbe","visdat":{"4ce1b576dbe":["function (y) ","x"],"4ce118e5dac3":["function (y) ","x"],"4ce11a7d44fa":["function (y) ","x"],"4ce1163b86ac":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div style="width: 50%">
<div id="htmlwidget-29c02916079b754882b9" style="width:100%;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-29c02916079b754882b9">{"x":{"data":[{"x":[-1.5,0.5],"y":[1,0],"text":"n1b -> n2b","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(124,174,0,1)","dash":"dot"},"hoveron":"points","name":"(admix,1)","legendgroup":"(admix,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,0.5],"y":[2,0],"text":"n2 -> n2b","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"dot"},"hoveron":"points","name":"(admix,2)","legendgroup":"(admix,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.5,0.5],"y":[0,-1],"text":"n2b -> R1","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"(normal,0)","legendgroup":"(normal,0)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5,-1.5],"y":[1,0],"text":"n1b -> L2","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(124,174,0,1)","dash":"solid"},"hoveron":"points","name":"(normal,1)","legendgroup":"(normal,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1,-1.5,null,-1,-0.5,null,1,1.5],"y":[2,1,null,2,1,null,2,1],"text":["n1 -> n1b","n1 -> n1b",null,"n1 -> L1","n1 -> L1",null,"n2 -> R2","n2 -> R2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)","dash":"solid"},"hoveron":"points","name":"(normal,2)","legendgroup":"(normal,2)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,null,0,1],"y":[3,2,null,3,2],"text":["R -> n1","R -> n1",null,"R -> n2","R -> n2"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(199,124,255,1)","dash":"solid"},"hoveron":"points","name":"(normal,3)","legendgroup":"(normal,3)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,-1,-1.5,0.5,1],"y":[3,2,1,0,2],"text":["R","n1","n1b","n2b","n2"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":0,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,0.5,-1.25,-1.5,-0.5,-0.75,0.5,1.25,0.75],"y":[2.5,2.5,1.5,0.5,0.5,1.5,-0.5,1.5,1],"text":["","","","","","","","",""],"hovertext":["R -> n1","R -> n2","n1 -> n1b","n1b -> L2","n1b -> n2b","n1 -> L1","n2b -> R1","n2 -> R2","n2 -> n2b"],"textfont":{"size":9.4488188976378,"color":"rgba(0,0,0,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.5],"y":[-1.1],"text":"R1","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(248,118,109,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(0,1)","legendgroup":"(0,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-1.5],"y":[-0.1],"text":"L2","hovertext":"","textfont":{"size":9.4488188976378,"color":"rgba(124,174,0,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"1","legendgroup":"1","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[-0.5,1.5],"y":[0.9,0.9],"text":["L1","R2"],"hovertext":["",""],"textfont":{"size":9.4488188976378,"color":"rgba(0,191,196,1)"},"type":"scatter","mode":"text","hoveron":"points","name":"(2,1)","legendgroup":"(2,1)","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":27.8246364188244,"r":7.30593607305936,"b":15.4776044553541,"l":10.958904109589},"paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.9,1.9],"tickmode":"array","ticktext":["-1","0","1"],"tickvals":[-1,0,1],"categoryorder":"array","categoryarray":["-1","0","1"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1.305,3.205],"tickmode":"array","ticktext":["-1","0","1","2","3"],"tickvals":[-1,0,1,2,3],"categoryorder":"array","categoryarray":["-1","0","1","2","3"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.65296803652968,"tickwidth":0,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(255,255,255,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"4ce17154ca92":{"linetype":{},"colour":{},"text":{},"x":{},"y":{},"xend":{},"yend":{},"from":{},"to":{},"type":"scatter"},"4ce175bf715e":{"x":{},"y":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"4ce16bac0eb6":{"x":{},"y":{},"label":{},"text":{},"x.1":{},"y.1":{},"xend":{},"yend":{},"from":{},"to":{}},"4ce16cfa05f3":{"label":{},"colour":{},"from":{},"x":{},"y":{},"xend":{},"yend":{},"from.1":{},"to":{}}},"cur_data":"4ce17154ca92","visdat":{"4ce17154ca92":["function (y) ","x"],"4ce175bf715e":["function (y) ","x"],"4ce16bac0eb6":["function (y) ","x"],"4ce16cfa05f3":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p><em>qpWave</em> allows us to estimate a lower bound on the number of gene flow events that separate two groups of populations <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> of size <span class="math inline">\(n_L\)</span> and <span class="math inline">\(n_R\)</span>.</p>
<p>The basic idea behind it is that we can use <span class="math inline">\(f_4\)</span>-statistics of the form <span class="math inline">\(f_4(L_i, L_j; R_m, R_n)\)</span> to estimate the genetic drift that is shared between pairs of left populations and pairs of right populations. There are <span class="math inline">\(\frac{n_L(n_L-1)}{2} \frac{n_R(n_R-1)}{2}\)</span> <span class="math inline">\(f_4\)</span>-statistics of that form, but not all of them contain unique information. In the case where <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> form two clades, all <span class="math inline">\(f_4\)</span>-statistics <span class="math inline">\(f_4(L_i, L_j; R_m, R_n)\)</span> are expected to be zero. If there are two independent gene flow events, a few of these <span class="math inline">\(f_4\)</span>-statistics will be enough to determine what the other <span class="math inline">\(f_4\)</span>-statistics should be. More generally, we can construct a matrix <span class="math inline">\(X\)</span> of these <span class="math inline">\(f_4\)</span>-statistics, and the rank of <span class="math inline">\(X\)</span> tells us what the minimum number of gene flows is between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>.</p>
<p>In practice, <span class="math inline">\(X\)</span> is constructed by fixing one population in <span class="math inline">\(L\)</span> and one population in <span class="math inline">\(R\)</span> which results in a matrix of dimensions <span class="math inline">\((n_L - 1) \times (n_R - 1)\)</span> with all c of the form <span class="math inline">\(f_4(L_1, L_i; R_1, R_j)\)</span>. By convention, <span class="math inline">\(n_R &gt; n_L\)</span>, which means that the highest possible rank of <span class="math inline">\(X\)</span> is <span class="math inline">\(n_L - 1\)</span>. A rank of <span class="math inline">\(n_L - 1\)</span> suggests that at least <span class="math inline">\(n_L - 1\)</span> independent gene flows have occurred between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>.</p>
</div>
<div id="estimating-the-rank-of-x" class="section level3">
<h3 class="hasAnchor">
<a href="#estimating-the-rank-of-x" class="anchor"></a>Estimating the rank of X</h3>
<p>When estimating the rank of <span class="math inline">\(X\)</span>, we have to take into account that the <span class="math inline">\(f_4\)</span>-statistics are not estimated perfectly. If we didn’t take that into account, <span class="math inline">\(X\)</span> would always be of full rank.</p>
<p>To estimate the rank of <span class="math inline">\(X\)</span> while taking into account the noise in the estimates of <span class="math inline">\(f_4\)</span>, we test how well <span class="math inline">\(X\)</span> can be approximated by different low rank approximations. An approximation of <span class="math inline">\(X\)</span> with one rank less than full rank can be written as <span class="math inline">\(AB\)</span>, where <span class="math inline">\(A\)</span> is <span class="math inline">\((n_L-1) \times (n_L-2)\)</span>, and <span class="math inline">\(B\)</span> is <span class="math inline">\((n_L-2) \times (n_R-1)\)</span> and <span class="math inline">\(AB\)</span> should be as close as possible to <span class="math inline">\(X\)</span>. In this case (one rank less than full rank), <span class="math inline">\(X\)</span> would have rank <span class="math inline">\(n_L - 1\)</span>, and <span class="math inline">\(AB\)</span> would have rank <span class="math inline">\(n_L - 2\)</span>. If this approximation works well, the residuals <span class="math inline">\(E = X - AB\)</span> will be very small. The tricky part here is that we need to take into account that the <span class="math inline">\(f_4\)</span>-statistics are estimated with some amount noise, and that they are highly correlated to each other. To account for that, we</p>
<p>The difference of these two log likelihoods approximately follows a <span class="math display">\[ \Chi^2 \]</span> distribution with one degree of freedom,</p>
<p>To formally test how much information is added by including T in L, <span class="math display">\[ X' \]</span> is compared to a low rank approximation of itself in which the rank has an upper bound of <span class="math display">\[ n_L \]</span>. This low rank approximation of <span class="math display">\[ X' \]</span> can be written as <span class="math display">\[ AB \]</span>, where <span class="math display">\[ A \]</span> is <span class="math display">\[ n_L \times n_L \]</span>, and <span class="math display">\[ B \]</span> is <span class="math display">\[ n_L \times n_R \]</span> and should minimize the log likelihood <span class="math display">\[ L(A,B) = -\frac{1}{2} E'Q^{-1}E  \]</span>. Here, <span class="math display">\[ E \]</span> is a <span class="math display">\[ n_L (n_R-1) \]</span> vector of residuals given by <span class="math display">\[ E = X' - AB \]</span>, and <span class="math display">\[ Q \]</span> is the <span class="math display">\[ n_L (n_R-1) \times n_L (n_R-1) \]</span> covariance matrix of all pairs of f4-statistics in <span class="math display">\[ X' \]</span>, estimated using block jackknife. An initial estimate of <span class="math display">\[ AB \]</span> is obtained through a singular value decomposition (SVD) of <span class="math display">\[ X' \]</span>. If the errors in the estimated f4-statistics were all uncorrelated, this singular value decomposition (SVD) of <span class="math display">\[ X' \]</span> would already yield <span class="math display">\[ A \]</span> and <span class="math display">\[ B \]</span> which minimizes <span class="math display">\[ L(A, B) \]</span>. However, since this is generally not the case, <span class="math display">\[ A \]</span> and <span class="math display">\[ B \]</span> are further optimized in an iterative procedure where one matrix is held constant, while the other matrix is adjusted to minimize <span class="math display">\[ L(A, B) \]</span>. is compared to a low rank approximation of itself in which the rank has an upper bound of . This low rank approximation of can be written as , where is , and is and should minimize the log likelihood . Here, is a vector of residuals given by , and is the covariance matrix of all pairs of f4-statistics in , estimated using block jackknife. An initial estimate of is obtained through a singular value decomposition (SVD) of . If the errors in the estimated f4-statistics were all uncorrelated, this singular value decomposition (SVD) of would already yield and which minimizes . However, since this is generally not the case, and are further optimized in an iterative procedure where one matrix is held constant, while the other matrix is adjusted to minimize . This likelihood is then compared to the likelihood of the full model, where <span class="math display">\[ X' \]</span> is decomposed into two matrices without constraining the rank to be <span class="math display">\[ n_L \]</span>. The difference of these two log likelihoods approximately follows a <span class="math display">\[ \Chi^2 \]</span> distribution with one degree of freedom, and this makes it possible to test whether the populations in L fully explain the variation in T, relative to R. The more general case where rank differences can be greater than one is described in [cite haak]. is decomposed into two matrices without constraining the rank to be . The difference of these two log likelihoods approximately follows a distribution with one degree of freedom, and this makes it possible to test whether the populations in L fully explain the variation in T, relative to R. The more general case where rank differences can be greater than one is described in [cite haak]. The admixture weights for the populations in L relative to T are given by left null space of the matrix <span class="math display">\[ A \]</span>, normalized so that they sum to one., normalized so that they sum to one.</p>
</div>
</div>
<div id="qpadm" class="section level2">
<h2 class="hasAnchor">
<a href="#qpadm" class="anchor"></a>qpAdm</h2>
<div id="modelling-an-admixed-population" class="section level3">
<h3 class="hasAnchor">
<a href="#modelling-an-admixed-population" class="anchor"></a>Modelling an admixed population</h3>
<p><em>qpAdm</em> extends <em>qpWave</em> by adding an additional target population (<span class="math inline">\(T\)</span>) to the left and right population sets <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>, in order to assess whether <span class="math inline">\(T\)</span> can be modeled as a mixture of populations in <span class="math inline">\(L\)</span>, at the exclusion of gene flow from populations in <span class="math inline">\(R\)</span>. The null hypothesis tested by <em>qpAdm</em> is that the populations in <span class="math inline">\(L\)</span> can account for all shared drift between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span>. A significant p-value rejecting this null hypothesis indicates that there has been gene flow between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span> (in either direction) that is not accounted for by any population in <span class="math inline">\(L\)</span>.</p>
<p>Often the goal of <em>qpAdm</em> modeling is to identify a set of populations <span class="math inline">\(L\)</span> for which this null hypothesis does not fail (i.e. there was no gene flow between <span class="math inline">\(R\)</span> and <span class="math inline">\(T\)</span>, relative to <span class="math inline">\(L\)</span>), and to estimate the relative contribution of the populations in <span class="math inline">\(L\)</span> to <span class="math inline">\(T\)</span>. If no set of populations <span class="math inline">\(L\)</span> can be found which fully explain the variation in <span class="math inline">\(T\)</span> relative to <span class="math inline">\(R\)</span> (all models fail), this may hint at the existence of a yet undiscovered “ghost” population which is related to a population in <span class="math inline">\(R\)</span>, but not represented by any of the populations in <span class="math inline">\(L\)</span>. If a set of populations <span class="math inline">\(L\)</span> can be found which fully explains the variation in <span class="math inline">\(T\)</span> relative to <span class="math inline">\(R\)</span>, <span class="math inline">\(T\)</span> can be thought of as a mixture of source populations which are represented by populations in <span class="math inline">\(L\)</span>, and <em>qpAdm</em> will estimate the mixture proportions. However, it is important to keep in mind that the true source populations are almost certainly not identical to any populations in <span class="math inline">\(L\)</span>, and that any inferences about the mixture proportions of the true, unobserved sources depends on an unverifiable assumption: Each left population needs to be strictly cladal with one of the true, unobserved source populations.</p>
</div>
<div id="how-it-works" class="section level3">
<h3 class="hasAnchor">
<a href="#how-it-works" class="anchor"></a>How it works</h3>
<p>Briefly, <em>qpAdm</em> is based on comparing to matrices of <span class="math inline">\(f_4\)</span>-statistics: The first matrix captures shared drift between populations in <span class="math inline">\(L\)</span> and in <span class="math inline">\(R\)</span>, and the second matrix captures shared drift between populations in <span class="math inline">\(L \cup T\)</span> and populations in <span class="math inline">\(R\)</span>. If both matrices have full rank, it suggests that there was gene flow between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span> which does not pass through any populations in <span class="math inline">\(L\)</span>. In other words, <span class="math inline">\(T\)</span> cannot be modeled as a mix of populations in <span class="math inline">\(L\)</span>, without missing a component that is only shared between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span>.</p>
<p>The same in more words:</p>
<p>The starting point of <em>qpAdm</em> is the identification of sets of populations <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span> such that no pair of populations in <span class="math inline">\(L\)</span> forms a clade relative to all populations in <span class="math inline">\(R\)</span>. If that condition is met, the matrix of <span class="math inline">\(f_4\)</span>-statistics (<span class="math inline">\(X\)</span>) will have full rank. <span class="math inline">\(T\)</span> is then added to the populations in <span class="math inline">\(L\)</span>, which creates a new matrix of <span class="math inline">\(f_4\)</span>-statistics, <span class="math inline">\(X'\)</span> of dimensions <span class="math inline">\(n_L \times (n_R - 1)\)</span>. If the rank of <span class="math inline">\(X'\)</span> is significantly higher than the rank of <span class="math inline">\(X\)</span>, this indicates that the new f4-statistics contain information about shared drift which was not already present in the old <span class="math inline">\(f_4\)</span>-statistics, and that gene flow has occurred between <span class="math inline">\(R\)</span> and <span class="math inline">\(T\)</span> which is not captured by any population in <span class="math inline">\(L\)</span>. On the other hand, if the rank of <span class="math inline">\(X'\)</span> does not increase significantly, it suggests that <span class="math inline">\(T\)</span> does not share any drift with <span class="math inline">\(R\)</span> which is not captured by <span class="math inline">\(L\)</span> (there are no paths from <span class="math inline">\(T\)</span> to any population in <span class="math inline">\(R\)</span> which do not pass through ancestors of a population in <span class="math inline">\(L\)</span>). This implies that all shared drift between <span class="math inline">\(T\)</span> and <span class="math inline">\(R\)</span> can be modeled as a linear combination of shared drift between <span class="math inline">\(L\)</span> and <span class="math inline">\(R\)</span>.</p>
<div id="estimating-admixture-proportions" class="section level4">
<h4 class="hasAnchor">
<a href="#estimating-admixture-proportions" class="anchor"></a>Estimating admixture proportions</h4>
<p>If the condition mentioned above is met (<span class="math inline">\(f_4\)</span>-statistics involving <span class="math inline">\(T\)</span> can be modeled as a linear combination of <span class="math inline">\(f_4\)</span>-statistics without <span class="math inline">\(T\)</span>), then the coefficients of this linear combination correspond to the relative contributions of the source populations.</p>
</div>
<div id="testing-rank-differences" class="section level4">
<h4 class="hasAnchor">
<a href="#testing-rank-differences" class="anchor"></a>Testing rank differences</h4>
<p>The weights of the linear combination of <span class="math inline">\(f_4\)</span>-statistics excluding <span class="math inline">\(T\)</span> which can correspond to the relative contributions of the source populations. The weights <span class="math inline">\(w\)</span> are normalized so that they sum to one, and they are equivalent to the left null space of the matrix <span class="math inline">\(A\)</span> (they satisfy <span class="math inline">\(w A = 0\)</span>), where <span class="math inline">\(A B\)</span> is a rank <span class="math inline">\(n_L - 1\)</span> approximation of <span class="math inline">\(X'\)</span>. will have full rank. T is then added to the populations in L, forming a new matrix of f4-statistics, of dimensions . If the rank of is significantly higher than the rank of , this indicates that the new f4-statistics contain information about shared drift which was not already present in the old f4-statistics, and that gene flow has occurred between R and T which is not captured by any population in L. On the other hand, if the rank of does not increase significantly, it suggests that T does not share any drift with R which is not captured by L (there are no paths from T to any population in R which do not pass through ancestors of a population in L). This implies that all shared drift between T and R can be modeled as a linear combination of shared drift between L and R. If T is admixed between source populations which form clades with populations in L, the weights of this linear combination will correspond to the relative contributions of the source populations. The weights are normalized so that they sum to one, and they are equivalent to the left null space of the matrix (they satisfy ), where is a rank approximation of .</p>
<p>Whereas the purpose of qpGraph is to model the ancestral relationships among a set of populations without paying special attention to a particular population, the purpose of qpAdm and qpWave is to test whether a target population T can be modeled as a mixture of a set of left populations L (proxy source populations), relative to a set of more distantly related right populations R. No assumptions are made about the relationships among the populations in L and R. There are simpler methods which can model a population as a mixture of other populations. For example, variations of PCA or ADMIXTURE can model allele frequencies in one population as a linear combination of allele frequencies in other populations. These methods do not require a set of right populations, so why are they necessary in the qpAdm framework? The right populations serve two purposes: First, they allow us to test whether the L is sufficient to explain all variation in T, or whether there is evidence of shared drift between T and R which is not present in L. Second, they make it possible to assign the correct weights to populations in L even in cases where the populations in L have a complex demographic history. For this it is necessary that the populations in R are differentially related to the populations in L. In most cases the sampled left populations L will not be the true source populations for T. However, the ancestry proportions of the true source populations can still be estimated accurately if each true source population forms a clade with its representative in L relative to all other modelled populations, and if the right populations are chosen such that no two left populations form a clade relative to all right populations. While the second assumption can be tested, the first assumption is untestable because it depends on potentially unobserved populations, which should be considered when interpreting qpAdm results. The null hypothesis tested by qpAdm is that the populations in L can account for all shared drift between T and R. A significant p-value rejecting this null hypothesis indicates that there has been gene flow between T and R (in either direction) that is not accounted for by any population in L. Often the goal of qpAdm modelling is to identify a set of populations L for which the null hypothesis doesn’t fail, and to estimate the relative contribution of each of these populations to T. However, it’s important to keep in mind that a qpAdm model with a non-significant p-value is not necessarily correct. There may be insufficient power to reject an incorrect model for a variety of reasons. If no set of populations L can be found which fully explain the variation in T relative to R (all models fail), this may hint at the existence of a yet undiscovered “ghost” population which is related to a population in R, but not represented by any of the populations in L. qpWave uses the same framework as qpAdm to estimate the number of independent ancestry sources which have contributed variation to T.</p>
<p>qpAdm works by constructing a matrix of f4-statistics between populations in L and T on one side, and populations in R on the other side. If this matrix has full rank, it suggests that each population in L and T has unique shared drift with R, and so the model fails: L alone cannot explain all shared drift between R and T.</p>
<p>In more detail, qpAdm first considers all f4-statistics of the form <span class="math display">\[ f4(L_1, L_i; R_1, R_j) \]</span>, where <span class="math display">\[ i \]</span> and <span class="math display">\[ j \]</span> index all populations in L and R, respectively, except for <span class="math display">\[ L_1 \]</span> and <span class="math display">\[ R_1 \]</span>. This <span class="math display">\[ (n_L-1) \times (n_R-1) \]</span> matrix <span class="math display">\[ X \]</span> of f4-statistics has full rank (all rows are linearly independent), only if all populations in L are differentially related to one or more populations in R. In other words, if no two populations in L form a clade with respect to all populations in R. Otherwise, if the rank of <span class="math display">\[ X \]</span> is less than <span class="math display">\[ n_L-1 \]</span>, it suggests that at least one row of <span class="math display">\[ X \]</span> is redundant because there are two populations in L which form a clade with respect to all populations in R., where and index all populations in L and R, respectively, except for and . This matrix of f4-statistics has full rank (all rows are linearly independent), only if all populations in L are differentially related to one or more populations in R. In other words, if no two populations in L form a clade with respect to all populations in R. Otherwise, if the rank of is less than , it suggests that at least one row of is redundant because there are two populations in L which form a clade with respect to all populations in R. Once a combination of L and R has been found so that <span class="math display">\[ X \]</span> is full rank, we add T to L and compute a new matrix <span class="math display">\[ X' \]</span> of dimensions <span class="math display">\[ (n_L+1) \times n_R \]</span>. If the rank of <span class="math display">\[ X' \]</span> is equal to the rank of <span class="math display">\[ X \]</span>, the inclusion of T did not add any new information relative to R: The variation in T can be fully explained by L. On the other hand, if the rank of X’ increases by one relative to the rank of <span class="math display">\[ X' \]</span>, it means that there is shared drift between T and some population in R which is not captured by any of the pairs in L, and thus T cannot be adequately modelled as a mix of populations in L. is full rank, we add T to L and compute a new matrix of dimensions . If the rank of is equal to the rank of , the inclusion of T did not add any new information relative to R: The variation in T can be fully explained by L. On the other hand, if the rank of X’ increases by one relative to the rank of , it means that there is shared drift between T and some population in R which is not captured by any of the pairs in L, and thus T cannot be adequately modelled as a mix of populations in L. To formally test how much information is added by including T in L, <span class="math display">\[ X' \]</span> is compared to a low rank approximation of itself in which the rank has an upper bound of <span class="math display">\[ n_L \]</span>. This low rank approximation of <span class="math display">\[ X' \]</span> can be written as <span class="math display">\[ AB \]</span>, where <span class="math display">\[ A \]</span> is <span class="math display">\[ n_L \times n_L \]</span>, and <span class="math display">\[ B \]</span> is <span class="math display">\[ n_L \times n_R \]</span> and should minimize the log likelihood <span class="math display">\[ L(A,B) = -\frac{1}{2} E'Q^{-1}E  \]</span>. Here, <span class="math display">\[ E \]</span> is a <span class="math display">\[ n_L (n_R-1) \]</span> vector of residuals given by <span class="math display">\[ E = X' - AB \]</span>, and <span class="math display">\[ Q \]</span> is the <span class="math display">\[ n_L (n_R-1) \times n_L (n_R-1) \]</span> covariance matrix of all pairs of f4-statistics in <span class="math display">\[ X' \]</span>, estimated using block jackknife. An initial estimate of <span class="math display">\[ AB \]</span> is obtained through a singular value decomposition (SVD) of <span class="math display">\[ X' \]</span>. If the errors in the estimated f4-statistics were all uncorrelated, this singular value decomposition (SVD) of <span class="math display">\[ X' \]</span> would already yield <span class="math display">\[ A \]</span> and <span class="math display">\[ B \]</span> which minimizes <span class="math display">\[ L(A, B) \]</span>. However, since this is generally not the case, <span class="math display">\[ A \]</span> and <span class="math display">\[ B \]</span> are further optimized in an iterative procedure where one matrix is held constant, while the other matrix is adjusted to minimize <span class="math display">\[ L(A, B) \]</span>. is compared to a low rank approximation of itself in which the rank has an upper bound of . This low rank approximation of can be written as , where is , and is and should minimize the log likelihood . Here, is a vector of residuals given by , and is the covariance matrix of all pairs of f4-statistics in , estimated using block jackknife. An initial estimate of is obtained through a singular value decomposition (SVD) of . If the errors in the estimated f4-statistics were all uncorrelated, this singular value decomposition (SVD) of would already yield and which minimizes . However, since this is generally not the case, and are further optimized in an iterative procedure where one matrix is held constant, while the other matrix is adjusted to minimize . This likelihood is then compared to the likelihood of the full model, where <span class="math display">\[ X' \]</span> is decomposed into two matrices without constraining the rank to be <span class="math display">\[ n_L \]</span>. The difference of these two log likelihoods approximately follows a <span class="math display">\[ \Chi^2 \]</span> distribution with one degree of freedom, and this makes it possible to test whether the populations in L fully explain the variation in T, relative to R. The more general case where rank differences can be greater than one is described in [cite haak]. is decomposed into two matrices without constraining the rank to be . The difference of these two log likelihoods approximately follows a distribution with one degree of freedom, and this makes it possible to test whether the populations in L fully explain the variation in T, relative to R. The more general case where rank differences can be greater than one is described in [cite haak]. The admixture weights for the populations in L relative to T are given by left null space of the matrix <span class="math display">\[ A \]</span>, normalized so that they sum to one., normalized so that they sum to one.</p>
</div>
</div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'c30b6be96312414d83a0ba52b94b558b',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
