<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ admixtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>


  
  <script src="../extra.js"></script>
<meta property="og:title" content="Function reference" />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-155330476-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-155330476-1');
</script>


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/admixtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fstats.html">Computing *f*-statistics</a>
    </li>
    <li>
      <a href="../articles/plotting.html">Plotting</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-admixtools-programs" class="hasAnchor"><a href="#section-admixtools-programs" class="anchor"></a>Admixtools programs</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="qp3pop.html">qp3pop()</a></code> </p>
        </td>
        <td><p>Estimate f3 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat.html">qpdstat()</a></code> </p>
        </td>
        <td><p>Estimate f4 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm.html">qpadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph.html">qpgraph()</a></code> </p>
        </td>
        <td><p>Compute the fit of an admixture graph</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-wrapper-functions" class="hasAnchor"><a href="#section-wrapper-functions" class="anchor"></a>Wrapper functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="qp3pop_wrapper.html">qp3pop_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qp3Pop program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_wrapper.html">qpadm_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpAdm program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat_wrapper.html">qpdstat_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpDstat program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_wrapper.html">qpgraph_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpGraph program</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-parsing-functions" class="hasAnchor"><a href="#section-parsing-functions" class="anchor"></a>Parsing functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="parse_qp3pop_output.html">parse_qp3pop_output()</a></code> </p>
        </td>
        <td><p>Read qp3Pop output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpadm_output.html">parse_qpadm_output()</a></code> </p>
        </td>
        <td><p>Read qpAdm output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpdstat_output.html">parse_qpdstat_output()</a></code> </p>
        </td>
        <td><p>Read qpDstat output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_graphfile.html">parse_qpgraph_graphfile()</a></code> </p>
        </td>
        <td><p>Read qpGraph graph file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_output.html">parse_qpgraph_output()</a></code> </p>
        </td>
        <td><p>Read qpGraph output file</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plotting-functions" class="hasAnchor"><a href="#section-plotting-functions" class="anchor"></a>Plotting functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="plot_comparison.html">plot_comparison()</a></code> </p>
        </td>
        <td><p>Plot a comparison of two runs of qpgraph or two runs of qpadm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph.html">plot_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph_map.html">plot_graph_map()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Plot samples on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_graph.html">plotly_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph using plotly</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-resampling-functions" class="hasAnchor"><a href="#section-resampling-functions" class="anchor"></a>Resampling functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="resample_inds.html">qp3pop_resample_inds()</a></code> <code><a href="resample_inds.html">qpdstat_resample_inds()</a></code> <code><a href="resample_inds.html">qpadm_resample_inds()</a></code> <code><a href="resample_inds.html">qpgraph_resample_inds()</a></code> </p>
        </td>
        <td><p>Run models while leaving out individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_snps.html">qp3pop_resample_snps()</a></code> <code><a href="resample_snps.html">qpdstat_resample_snps()</a></code> <code><a href="resample_snps.html">qpadm_resample_snps()</a></code> <code><a href="resample_snps.html">qpgraph_resample_snps()</a></code> </p>
        </td>
        <td><p>Run models while leaving out SNP blocks</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-reading-and-writing" class="hasAnchor"><a href="#section-reading-and-writing" class="anchor"></a>Reading and writing</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="extract_f2.html">extract_f2()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_counts.html">extract_counts()</a></code> </p>
        </td>
        <td><p>Extract and store data needed to compute blocked f2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_precomp.html">f2_from_precomp()</a></code> </p>
        </td>
        <td><p>Read blocked f2 statistics from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_geno.html">f2_from_geno()</a></code> </p>
        </td>
        <td><p>Read f2 jackknife blocks from genotype data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_packedancestrymap.html">read_packedancestrymap()</a></code> </p>
        </td>
        <td><p>Read genotype data from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_aftable.html">packedancestrymap_to_aftable()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from packedancestrymap files</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-graph-topology" class="hasAnchor"><a href="#section-graph-topology" class="anchor"></a>Graph topology</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="find_graphs.html">find_graphs()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_admixturegraph.html">random_admixturegraph()</a></code> </p>
        </td>
        <td><p>Generate a random admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_triples.html">summarize_triples()</a></code> </p>
        </td>
        <td><p>Summarize triples across graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_models.html">qpadm_models()</a></code> </p>
        </td>
        <td><p>Return all valid qpAdm models for an admixturegraph</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-everything" class="hasAnchor"><a href="#section-everything" class="anchor"></a>Everything</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="admixtools.html">admixtools</a></code> </p>
        </td>
        <td><p>Lightweight R implementations of some AdmixTools programs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="afs_to_f2_blocks.html">afs_to_f2_blocks()</a></code> </p>
        </td>
        <td><p>Compute all pairwise f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ancestrymap_to_aftable.html">ancestrymap_to_aftable()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from ancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits2.html">compare_fits2()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits3.html">compare_fits3()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_fits4.html">compare_fits4()</a></code> </p>
        </td>
        <td><p>Compare the fit of two qpgraph models</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="decompose_graph.html">decompose_graph()</a></code> </p>
        </td>
        <td><p>Find all trees which are part of the admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="delete_groups.html">delete_groups()</a></code> </p>
        </td>
        <td><p>Delete groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="desimplify_graph.html">desimplify_graph()</a></code> </p>
        </td>
        <td><p>Add two nodes before each admixture node</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_afs.html">extract_afs()</a></code> </p>
        </td>
        <td><p>Compute and store blocked allele frequency data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_counts.html">extract_counts()</a></code> </p>
        </td>
        <td><p>Extract and store data needed to compute blocked f2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_f2.html">extract_f2()</a></code> </p>
        </td>
        <td><p>Compute and store blocked f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_f2_subset.html">extract_f2_subset()</a></code> </p>
        </td>
        <td><p>Copy a subset of f2-statistics to a new directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_samples.html">extract_samples()</a></code> </p>
        </td>
        <td><p>Extract samples from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2.html">f2()</a></code> </p>
        </td>
        <td><p>Estimate f2 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_geno.html">f2_from_geno()</a></code> </p>
        </td>
        <td><p>Read f2 jackknife blocks from genotype data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="f2_from_precomp.html">f2_from_precomp()</a></code> </p>
        </td>
        <td><p>Read blocked f2 statistics from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="find_graphs.html">find_graphs()</a></code> </p>
        </td>
        <td><p>Find well fitting admixture graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_block_lengths.html">get_block_lengths()</a></code> </p>
        </td>
        <td><p>Find LD-independent blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_flipadmix.html">graph_flipadmix()</a></code> </p>
        </td>
        <td><p>Find all valid graphs which result from flipping one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_minusone.html">graph_minusone()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from removing one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_minusplus.html">graph_minusplus()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from adding and removing one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graph_plusone.html">graph_plusone()</a></code> </p>
        </td>
        <td><p>Find all graphs which result from adding one admixture edge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="graphmod_pavel.html">graphmod_pavel()</a></code> </p>
        </td>
        <td><p>Return all graphs created from permuting a subclade</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_samples.html">group_samples()</a></code> </p>
        </td>
        <td><p>Group precomputed data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lazadm.html">lazadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_resample_inds_fun.html">make_resample_inds_fun()</a></code> </p>
        </td>
        <td><p>Takes a function <code>qpfun</code> which takes f2_blocks as input
Returns a function which will repeadetly evaluate <code>qpfun</code> on
a subset of all samples, returning a nested data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="make_resample_snps_fun.html">make_resample_snps_fun()</a></code> </p>
        </td>
        <td><p>Takes a function <code>qpfun</code> which takes f2_blocks as input
Returns a function which will repeadetly evaluate <code>qpfun</code> on
jackknife or bootstrap resamplings of the f2_blocks, returning a nested data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="msprime_sim.html">msprime_sim()</a></code> </p>
        </td>
        <td><p>Generate msprime simulation code from admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="newick_to_edges.html">newick_to_edges()</a></code> </p>
        </td>
        <td><p>Turn a newick format tree to a matrix of edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="node_counts.html">node_counts()</a></code> </p>
        </td>
        <td><p>Count how often each node in graph occurs in other graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="node_signature.html">node_signature()</a></code> </p>
        </td>
        <td><p>Returns a signature of a graph consisting of the left and right descendent leaf nodes of each internal node (sorted and concatenated)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="packedancestrymap_to_aftable.html">packedancestrymap_to_aftable()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qp3pop_output.html">parse_qp3pop_output()</a></code> </p>
        </td>
        <td><p>Read qp3Pop output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpadm_output.html">parse_qpadm_output()</a></code> </p>
        </td>
        <td><p>Read qpAdm output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpdstat_output.html">parse_qpdstat_output()</a></code> </p>
        </td>
        <td><p>Read qpDstat output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_graphfile.html">parse_qpgraph_graphfile()</a></code> </p>
        </td>
        <td><p>Read qpGraph graph file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parse_qpgraph_output.html">parse_qpgraph_output()</a></code> </p>
        </td>
        <td><p>Read qpGraph output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plink_to_aftable.html">plink_to_aftable()</a></code> </p>
        </td>
        <td><p>Read allele frequencies from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_comparison.html">plot_comparison()</a></code> </p>
        </td>
        <td><p>Plot a comparison of two runs of qpgraph or two runs of qpadm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph.html">plot_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_graph_map.html">plot_graph_map()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_map.html">plot_map()</a></code> </p>
        </td>
        <td><p>Plot samples on a map</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotly_graph.html">plotly_graph()</a></code> </p>
        </td>
        <td><p>Plot an admixture graph using plotly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qp3pop.html">qp3pop()</a></code> </p>
        </td>
        <td><p>Estimate f3 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qp3pop_wrapper.html">qp3pop_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qp3Pop program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm.html">qpadm()</a></code> </p>
        </td>
        <td><p>Estimate admixture weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_models.html">qpadm_models()</a></code> </p>
        </td>
        <td><p>Return all valid qpAdm models for an admixturegraph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_pairs.html">qpadm_pairs()</a></code> </p>
        </td>
        <td><p>Compute all pairwise qpwave p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_rotate.html">qpadm_rotate()</a></code> </p>
        </td>
        <td><p>Compute all pairwise qpadm p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpadm_wrapper.html">qpadm_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpAdm program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat.html">qpdstat()</a></code> </p>
        </td>
        <td><p>Estimate f4 statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpdstat_wrapper.html">qpdstat_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpDstat program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph.html">qpgraph()</a></code> </p>
        </td>
        <td><p>Compute the fit of an admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_precompute_f3.html">qpgraph_precompute_f3()</a></code> </p>
        </td>
        <td><p>Compute f3-statistics from f2-statistics.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qpgraph_wrapper.html">qpgraph_wrapper()</a></code> </p>
        </td>
        <td><p>Wrapper function around the original qpGraph program</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_admixturegraph.html">random_admixturegraph()</a></code> </p>
        </td>
        <td><p>Generate a random admixture graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_newick.html">random_newick()</a></code> </p>
        </td>
        <td><p>Generate a random, binary graph with n terminal nodes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_ancestrymap.html">read_ancestrymap()</a></code> </p>
        </td>
        <td><p>Read genotype data from ancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_f2.html">read_f2()</a></code> </p>
        </td>
        <td><p>Read blocked f2 estimates from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_packedancestrymap.html">read_packedancestrymap()</a></code> </p>
        </td>
        <td><p>Read genotype data from packedancestrymap files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_plink.html">read_plink()</a></code> </p>
        </td>
        <td><p>Read genotype data from <code>PLINK</code> files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_inds.html">qp3pop_resample_inds()</a></code> <code><a href="resample_inds.html">qpdstat_resample_inds()</a></code> <code><a href="resample_inds.html">qpadm_resample_inds()</a></code> <code><a href="resample_inds.html">qpgraph_resample_inds()</a></code> </p>
        </td>
        <td><p>Run models while leaving out individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resample_snps.html">qp3pop_resample_snps()</a></code> <code><a href="resample_snps.html">qpdstat_resample_snps()</a></code> <code><a href="resample_snps.html">qpadm_resample_snps()</a></code> <code><a href="resample_snps.html">qpgraph_resample_snps()</a></code> </p>
        </td>
        <td><p>Run models while leaving out SNP blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simplify_graph.html">simplify_graph()</a></code> </p>
        </td>
        <td><p>Remove redundant edges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_mat.html">split_mat()</a></code> </p>
        </td>
        <td><p>Split a matrix into blocks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarize_triples.html">summarize_triples()</a></code> </p>
        </td>
        <td><p>Summarize triples across graphs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="test_cladality.html">test_cladality()</a></code> </p>
        </td>
        <td><p>Test if two sets of populations form two clades</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_block_lengths.html">write_block_lengths()</a></code> </p>
        </td>
        <td><p>Compute and write block lengths</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_dot.html">write_dot()</a></code> </p>
        </td>
        <td><p>Convert graph to dot format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_f2.html">write_f2()</a></code> </p>
        </td>
        <td><p>Write blocked f2 estimates to disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_split_f2_block.html">write_split_f2_block()</a></code> </p>
        </td>
        <td><p>Compute f2 blocks and write them to disk</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-admixtools-programs">Admixtools programs</a></li>
      <li><a href="#section-wrapper-functions">Wrapper functions</a></li>
      <li><a href="#section-parsing-functions">Parsing functions</a></li>
      <li><a href="#section-plotting-functions">Plotting functions</a></li>
      <li><a href="#section-resampling-functions">Resampling functions</a></li>
      <li><a href="#section-reading-and-writing">Reading and writing</a></li>
      <li><a href="#section-graph-topology">Graph topology</a></li>
      <li><a href="#section-everything">Everything</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robert Maier, Nick Patterson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '',
    indexName: 'admixtools',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


